Fix: Selection does not follow mouse when scrolling and can show half-way between entries

vitvakatu opened the pull request:
### Pull Request Description

[ci no changelog needed]

[Task link](https://www.pivotaltracker.com/story/show/182713338).

This PR fixes the implementation of the entries highlight (selection box) in the component list panel of the component browser. The previous behavior was caused by the dumb implementation of keeping the selection inside the borders of the component list panel. Now the selection is correctly clipped and is hovering over entries at all times.

Screencasts:


https://user-images.githubusercontent.com/6566674/182196585-02ab5ec4-7d12-4d7a-8563-ac95aec3c9f4.mp4



https://user-images.githubusercontent.com/6566674/182196768-9d28e83d-6c77-4ef6-abb0-98d50c4e74b7.mp4


### Important Notes

- This PR does not change the fact that after opening the component browser, the selection hovers over a random entry. See a [dedicated task](#182729573) for that.
- The selection box shape was moved from the `component_group` module because it depends heavily on list panel view implementation.
- The selection was removed from the `component_group` demo scene. We have a `component_list_panel_view` demo scene instead.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide dist` and `./run ide watch`.

vitvakatu committed:
Improve implementation of the selection

farmaazon commented on app/gui/view/component-browser/component-group/src/lib.rs:
```diff
@@ -807,12 +790,16 @@ impl Model {
+            let half_header_height = header_geometry.height / 2.0;
+            let header_y = size.y / 2.0 - half_header_height - header_pos;
+            let header_y = header_y.max(-size.y / 2.0 + half_header_height);
+            Vector2(0.0, header_y)
```
I don't quite get the second point. We set the selection to the same position as header, and that's obvious. But why the second part? Does header pos is outside the component group sometimes?

vitvakatu replied:
I replaced this part of the code with a separate method. There is a case when we restricted the position of the header but didn't do it for the selection, so the selection was shifted beyond the group borders.

wdanilo commented on app/gui/view/component-browser/component-group/src/lib.rs:
```diff
@@ -807,12 +790,16 @@ impl Model {
+            // The following equations are similar to equations in the `resize` method.
+            // We make sure that selection position is
+            // - in the same place as the header position,
+            // - is not outside the component group.
```
- The sentence `We make sure that selection position is` ends without a dot or colon, so it's not valid.
- The points are neither sentences, nor do they end with semicolons, so it's not valid English.
- I don't really understand these docs. I understand what does it mean that "we make sure that ... is in the same place as ...", but there is no explanation of why it is needed, what happens if its not done, and how it is done (what does it mean that "we make sure").

vitvakatu replied:
These docs were removed after refactoring, and the part of the code is replaced by a dedicated method.

farmaazon commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -226,7 +233,72 @@ enum SectionHeaderPlacement {
+            let area_top = Rect((area_width.clone(),half_height.clone()));
+            let area_top = area_top.translate_y(quater_height.clone());
+            // Additional padding so that the two halves of the area are overlapping without gaps.
+            let padding = 20.0;
```
This is a magic number. Why 20.0 works? Would 10.0 also work? I think it could be rather a relative value, e.g. 1/8 of area size.

vitvakatu replied:
Fixed as proposed

farmaazon commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -226,7 +233,72 @@ enum SectionHeaderPlacement {
+            let area = area.translate_y(-margin_top.px() / 2.0);
+
+            let size = selection_size;
+            let mask = Rect((size.x().px(), size.y().px())).corners_radius(corners_radius.px());
```
IIRC you can just call Rect(size.px)

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -226,7 +233,72 @@ enum SectionHeaderPlacement {
+/// layers. See [component_group](component_group#selection) module documentation to learn how this
+/// mask is used.
+///
+/// The selection `mask` is a rectangle with rounded corners. The final shape is an intersection of
+/// the `mask` and the `area` shape that covers the entire component list panel and has the same
+/// shape. This is needed to ensure that the selection box never leaves the borders of the
```
I dont understand this sentence 

```
The final shape is an intersection of
  /// the `mask` and the `area` shape that covers the entire component list panel and has the same
  /// shape.
```

So, what has the same shape? The intersection and the component list?

vitvakatu replied:
The whole section is rewritten for clarity.

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -226,7 +233,72 @@ enum SectionHeaderPlacement {
+/// The selection `mask` is a rectangle with rounded corners. The final shape is an intersection of
+/// the `mask` and the `area` shape that covers the entire component list panel and has the same
+/// shape. This is needed to ensure that the selection box never leaves the borders of the
+/// component list panel. Instead of using `set_position` API methods of the display object, we
```
`This is needed to ensure that the selection box never leaves the borders of the component list panel` - sentence without explanation. Sure, we are ensuring something here, but what does it really mean? Why it is needed and why this is the best design? Should be explained.

vitvakatu replied:
The whole section is rewritten for clarity.

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -226,7 +233,72 @@ enum SectionHeaderPlacement {
+/// the `mask` and the `area` shape that covers the entire component list panel and has the same
+/// shape. This is needed to ensure that the selection box never leaves the borders of the
+/// component list panel. Instead of using `set_position` API methods of the display object, we
+/// have a `pos` parameter instead. Similarly, we have the `selection_size` parameter for the
```
- Please use a grammar checking tool. I was asking about it in previous PRs. Such sentences as `Instead of using `set_position` API methods of the display object, we have a `pos` parameter instead` are simply incorrect.
- Again, no explanation here. There is info that we dont use `set_position` API, but no explanation of why.

vitvakatu replied:
The whole section is rewritten for clarity.

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -226,7 +233,72 @@ enum SectionHeaderPlacement {
+/// have a `pos` parameter instead. Similarly, we have the `selection_size` parameter for the
+/// size of the `mask`.
+///
+/// The `area` shape consists of two parts and has a variable height. The top part is a simple
+/// rectangle, while the bottom part is a rounded rectangle, to repeat the rounded corners on the
+/// bottom of the component browser. The variable height is used to create an effect of the
```
I dont understand this sentence. What does it mean `to repeat the rounded corners on the bottom of the component browser`? Why are you repeating anything? Are you sure you used the correct verb?

vitvakatu replied:
The whole section is rewritten for clarity.

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -226,7 +233,72 @@ enum SectionHeaderPlacement {
+/// The `area` shape consists of two parts and has a variable height. The top part is a simple
+/// rectangle, while the bottom part is a rounded rectangle, to repeat the rounded corners on the
+/// bottom of the component browser. The variable height is used to create an effect of the
+/// covering the selection with the component group's header. When the header is selected,
```
"an effect of the covering the selection with the component group's header":
- not correct grammar.
- The "effect" is not explained. What does it mean?

vitvakatu replied:
The whole section is rewritten for clarity.

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -226,7 +233,72 @@ enum SectionHeaderPlacement {
+/// can't change the actual order of rendering here, but we clip the selection box so that it
+/// looks like the header covers the part of it.
+///
+/// We have only a single `corners_radius` parameter that is used both for the bottom corners of the
+/// `area` and the corners of the selection `mask` itself. In the current design, they are equal
```
- Why the word "itself" is needed here?
- "in the current design" - no link, no one knows what this sentence refers to.
- "because the header never touches the bottom of the component list panel and there is no way to see the difference in the radius of the corner." - no explanation agani. I don't understand what this sentence is trying to tell me.

vitvakatu replied:
The whole section is rewritten for clarity.

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -226,7 +233,72 @@ enum SectionHeaderPlacement {
+/// `area` and the corners of the selection `mask` itself. In the current design, they are equal
+/// when we select entries, they can be different when we choose the header. We don't need a
+/// separate parameter though, because the header never touches the bottom of the component list
+/// panel and there is no way to see the difference in the radius of the corner.
```
Overall, I find these docs in a pretty bad and not-explanatory shape. Please write them with proper explanation, and before creating PR, be sure to run all the docs trough some English grammar checker.

vitvakatu replied:
The whole section is rewritten for clarity.

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -846,15 +929,41 @@ impl component::Frp<Model> for Frp {
+            eval selection_size_animation.value ((pos) selection.selection_size.set(*pos));
             eval selection_corners_animation.value ((r) selection.corners_radius.set(*r));
             eval_ model.scroll_area.scroll_position_y(selection_animation.skip.emit(()));
+            // The following code changes the `margin_top` parameter of the selection shape to
+            // handle the cases where the selection is near the top of the component list panel.
+            // See the documentation of the [`selection_box`](selection_box) module for more
+            // details.
```
Every documentation should tell semantically what it does. It can redirect the reader to more detailed documentation, but it should at least shortly explain it. This doc does not explain things SEMANTICALLY. It tells what technically the follwoing code does, but without explanation of why. E.g. "handle the cases where the selection is near the top of the component list panel" - what cases? What is this even referring to?

vitvakatu replied:
Rewritten

farmaazon commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -797,17 +874,23 @@ impl component::Frp<Model> for Frp {
+        fn selection_position(
+            model: &Model,
+            id: GroupId,
+            group_local_pos: Vector2,
+            style: &Style,
+        ) -> Vector2 {
             let scroll_area = &model.scroll_area;
-            let scroll_area_pos = scroll_area.position() + scroll_area.content().position();
+            let scroll_area_size = style.scroll_area_size();
+            let scroll_area_center = Vector2(-scroll_area_size.x / 2.0, scroll_area_size.y / 2.0);
+            let scroll_area_pos = scroll_area_center + scroll_area.content().position().xy();
             let section_pos = match id.section {
                 SectionId::Favorites => model.favourites_section.content.position(),
                 SectionId::LocalScope => default(),
                 SectionId::SubModules => model.sub_modules_section.content.position(),
             };
             let group_pos = model.groups_wrapper.get(&id).map(|g| g.position()).unwrap_or_default();
-            let pos = (scroll_area_pos + section_pos + group_pos).xy() + group_local_pos;
-            model.clamp_y(pos)
+            scroll_area_pos + (section_pos + group_pos).xy() + group_local_pos
         }
```
I know it was not introdcuced in this PR, but can we make this function outside the function body?

vitvakatu replied:
Done.

vitvakatu committed:
Refactor docs

vitvakatu committed:
Refactoring component groups api

vitvakatu committed:
Polishing code

vitvakatu committed:
Merge branch 'develop' into wip/vitvakatu/proper-selection-182713338

# Conflicts:
#	app/gui/src/presenter/searcher.rs

vitvakatu committed:
Self-review

vitvakatu commented:
Status update:
I fixed a few additional issues and updated documentation, rewriting it for clarity and fixed grammar mistakes. I changed the API of the component groups - now, it correctly represents internal invariants using the type system (`is_header_selected` and `selected_entry` endpoints are merged into a single one). As a result, the selection works more reliably when selecting headers and scrolling the component list panel. See the updated videos in the PR description.

wdanilo commented on app/gui/src/presenter/searcher.rs:
```diff
@@ -280,15 +281,19 @@ impl Searcher {
-                    entry_selected <- list_view.selected_entry
-                        .map2(&list_view.is_header_selected, |entry, (_, is_header)| (!is_header).as_some(*entry))
-                        .filter_map(|e| *e);
+                    entry_selected <- list_view.selected.map(|s| match s {
+                        Some(list_panel::Selected::Entry(e)) => Some(*e),
+                        _ => None
+                    });
```
Linter should hint that this is the same as `.map`. Was the code linted? If so, could you check the linter configuration and see why it was not hinted here, please? The same applies to the code below. In either case, reply to this comment what happened here, please. https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0a8154155b3cb64fbc8b7378ce37a59b

vitvakatu replied:
Your example does not resemble the actual state of the code. [This one does](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8caa79f5e9a0f2cac2cedf4cac530205). This code uses the pattern matching to filter a specific enum variant, so it can't be expressed as `.map` call.

There is a way to rewrite it as `list_view.selected.map(|s| s.map(list_panel::Selected::as_entry))`, but it would require adding `as_entry` method which will do the similar pattern matching inside. I guess I can go that road.

vitvakatu replied:
Update: I implemented the `as_entry_id` method as part of the refactoring. The other similar match below doesn't worth it, I think. (or I somewhat can't find a good name for such a method)

wdanilo commented on app/gui/view/component-browser/component-group/src/lib.rs:
```diff
@@ -383,6 +366,14 @@ impl Colors {
 // === FRP ===
 // ===========
 
+/// The type of the [`selected`] FRP output of the component group.
```
This doc comment does not explain anything. It explains where this thing is being used but not WHAT it represents. Please write all comments explaining semantics of the data structure (what is the purpose of its existence). Please apply to all doc comments in this and future PRs.

vitvakatu replied:
Rephrased.

wdanilo commented on app/gui/view/component-browser/component-group/src/lib.rs:
```diff
@@ -807,15 +790,25 @@ impl Model {
         }
     }
 
+    /// The y position of the header. The position is restricted by the borders of the
+    /// component group.
```
Again, this doc comment does not really explains what this function does. It tells that it is "y position of the header", but this function contains a few lines of code that do something. But what they really do is not explained in the docs.

vitvakatu replied:
Rephrased, though I don't think the documentation must describe every implementation detail of every method. It is enough to know that this particular method returns a y position of the header, and this position is restricted by the borders of the component group.

wdanilo commented on app/gui/view/component-browser/component-group/src/set.rs:
```diff
@@ -136,10 +137,9 @@ propagated_events! {
     struct PropagatedEvents {
         mouse_in_group:            GroupId,
-        selected_entry:            Option<(GroupId, entry::Id)>,
+        selected:            Option<(GroupId, Selected)>,
```
alignment

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -226,7 +233,87 @@ enum SectionHeaderPlacement {
+/// [component_group](component_group#selection) documentation to learn more about how the component
+/// groups are implemented.
+///
+/// The selection box should never leave the borders of the component list panel. We can't use
+/// [layers masking][] as our renderer does not support hierarchical masks, so instead, we build a
+/// final shape as an intersection of the `mask` rounded rectangle and the `area` shape. The `area`
+/// covers the entire component list panel and thus limits the area in which the selection box is
+/// visible.
```
"so instead, we build a final shape as..." - a final shape of what?

vitvakatu replied:
Rephrased for clarity. The final shape of the selection box.

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -226,7 +233,87 @@ enum SectionHeaderPlacement {
+/// covers the entire component list panel and thus limits the area in which the selection box is
+/// visible.
+///
+/// The shape of the component list panel determines the size and the position of the `area`. One
```
The term `area` was introduced here but not explained.

vitvakatu replied:
I believe this is not true. It is introduced in two previous sentences.

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -226,7 +233,87 @@ enum SectionHeaderPlacement {
+/// visible.
+///
+/// The shape of the component list panel determines the size and the position of the `area`. One
+/// should use the display object's standard `set_position` API to move the `area`. The `mask` moves
```
```One should use the display object's standard `set_position` API to move the `area`. ``` - What is the purpose of this sentence? We can move all shapes using this API, so it should be obvious that we can move this shape as well, right? I feel I'm missing something here.

`The mask freely and represents the position of the selection box itself` - what does it mean? What does mean that it moves "freely"?

vitvakatu replied:
I rephrased this paragraph. We have two distinct ways to change the positions of two "components" of the selection box shape: the `mask` and the `area`.

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -226,7 +233,87 @@ enum SectionHeaderPlacement {
+///
+/// The `margin_top` parameter controls the variable height of the `area`. In a scrolled
+/// component group, the user can select a partially visible entry behind the group's header. We
+/// make the selection box appear covered by the selection box by reducing the height of the `area.`
```
```
We make the selection box appear covered by the selection box by reducing the height of the `area.`
```

What does it mean? Why selection box is covered by selection box?

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -226,7 +233,87 @@ enum SectionHeaderPlacement {
+/// The `margin_top` parameter controls the variable height of the `area`. In a scrolled
+/// component group, the user can select a partially visible entry behind the group's header. We
+/// make the selection box appear covered by the selection box by reducing the height of the `area.`
+/// It works because the header is always on top of the component list panel in a scrolled group.
```
I don't understand the sentence. Why selection box masked by selection box always work because header is on top?

vitvakatu replied:
That was a typo, thanks for catching it.

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -226,7 +233,87 @@ enum SectionHeaderPlacement {
+/// make the selection box appear covered by the selection box by reducing the height of the `area.`
+/// It works because the header is always on top of the component list panel in a scrolled group.
+///
+/// We have only a single `corners_radius` parameter for the bottom corners of the `area` and the
```
Then this should be named `bottom_corners_radius`

vitvakatu replied:
I'd somewhat disagree because the selection box has rounded corners on all sides, not only on the bottom.

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -526,6 +617,21 @@ impl Model {
         Vector2(pos.x, y)
     }
 
+    /// Calculate the view-local position of the selection from the component group-local one.
```
Could you expand these docs please? I was reading this sentence few times and I dont really understand what it does.

vitvakatu replied:
Expanded

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -760,6 +866,26 @@ impl EntryId {
     }
 }
 
+/// The selected part of the component group and the identifier of this group.
+///
```
If this structure is a "component group AND the identifier of this group" then it should contain BOTH information, but it contains only one of them.

vitvakatu replied:
Slightly refactored this struct. It contains group id in both variants, it just wasn't obvious enough.

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -760,6 +866,26 @@ impl EntryId {
+/// Similar to [`component_group::Selected`], but also contains a group id.
+#[allow(missing_docs)]
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq)]
+pub enum Selected {
```
This is defined in the FRP section - is it part of FRP? (Maybe it is, just asking)

vitvakatu replied:
Yes, this is a helper type for FRP API.

wdanilo commented on app/gui/view/component-browser/searcher-list-panel/src/lib.rs:
```diff
@@ -846,15 +959,42 @@ impl component::Frp<Model> for Frp {
+            eval selection_size_animation.value ((pos) selection.selection_size.set(*pos));
             eval selection_corners_animation.value ((r) selection.corners_radius.set(*r));
             eval_ model.scroll_area.scroll_position_y(selection_animation.skip.emit(()));
+            // When the selection highlights entries near the header of the component group, we use
+            // the `margin_top` parameter of the [`selection_box`] shape to clip the selection shape
+            // and make it appear covered by the header. When we select the header, the `margin_top`
+            // parameter is reset to zero to avoid clipping the selection shape.
+            // See the documentation of the [`selection_box`](selection_box) module for more
+            // details.
```
This is super cool doc - short, very clear, and explains things properly! <3

wdanilo commented on app/gui/view/debug_scene/component-group/src/lib.rs:
```diff
@@ -359,12 +354,14 @@ fn init(app: &Application) {
-        eval selected_entry([](e) if let Some((g, e)) = e {  DEBUG!("Entry {e} from group {g:?} selected") });
+        selected <- multiview.selected.on_change();
+        eval selected([](s) match s {
+            Some((g, Selected::Header)) => DEBUG!("Header selected in group {g:?}"),
+            Some((g, Selected::Entry(id))) => DEBUG!("Entry {id} from group {g:?} selected"),
+            _ => {},
```
debug left in the code

vitvakatu replied:
This is a debug scene that uses console messages for debugging purposes. This was not introduced in this PR and is only touched during refactoring.

vitvakatu committed:
Refactoring

vitvakatu committed:
Merge branch 'develop' into wip/vitvakatu/proper-selection-182713338

mergify[bot] committed:
Merge branch 'develop' into wip/vitvakatu/proper-selection-182713338
